### PR TITLE
Webgpu support

### LINKED ISSUE #1030: WebGPU render target?
Curiously with my GPU neither TensorFlow nor PyTorch work on Linux|Windows, but WebGPU with Chrome on Windows does. Will probably contribute to ROCm to add support, but in the interim it seems like a handy target for a project like tinygrad.

Of all tutorials, I found this one to be cleanest - https://webgpufundamentals.org/webgpu/lessons/webgpu-fundamentals.html

But I suppose the question becomes how to target WebGPU. Would one generate the JavaScript code directly, or with CPython soon getting WASM support should this be contributed to CPython and then tinygrad used to test this support on that CPython? - Or target C, as you do, and transform the output for WebGPU support (e.g., I could extend my cdd-c compiler project to support that; or could use LLVM's LibTooling).

PS: [offtopic] Oh also if you're interested in this support, maybe I'll enhance my (Apache-2.0 OR MIT) cdd-python compiler project to translate entire codebases to tinygrad. I'm thinking of taking PyTorch Lightning, torchvision and similar projects and auto-translating their calls to be tinygrad calls; and their imports to tinygrads'. Would be interesting from a regression testing perspective, but it should be doable without causing digressions from their master branches. I.e., you could automatically get all their loss functions, optimisers, models, data loaders &etc. [/offtopic]

Anyway; I digress. What do you/community think about WebGPU? - A good prospective target, or something to rely on other target backends to support?

### PR DESCRIPTION
**Pull request by dc-dc-dc**
In collab with @[martinloretzzz](https://github.com/geohot/tinygrad/commits?author=martinloretzzz)

Closes #1030

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/codegen/linearizer.py             3               3
tinygrad/jit.py                            1               1
tinygrad/runtime/ops_webgpu.py           133               0
------------------------------------------------------------
total                                    137               4
------------------------------------------------------------
lines added in the tinygrad folder: 133
```

File: tinygrad/runtime/ops_webgpu.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+    assert len([x for x in self.bufs if not isinstance(x, LocalBuffer) and not isinstance(x.realized, RawConst)]) <= 31, "WEBGPU max number of buffers is 31" 
+    function_name = f"{self.function_name}_{abs(hash(self.key))}"
+    bind_it = iter(range(len(self.bufs)))
+    prg = "\n".join([f"@group(0) @binding({next(bind_it)}) var<storage,read_write> data{i}: array<{type_map[x.dtype]}>;" for i,x in enumerate(self.bufs) if not isinstance(x, LocalBuffer) and not isinstance(x.realized, RawConst)])
+    prg += f"\n@compute @workgroup_size(1) fn {function_name}(@builtin(global_invocation_id) gindex: vec3<u32>, @builtin(local_invocation_id) lindex: vec3<u32>) {{\n" + "\n".join(kernel) + "\n}" # TODO: revert local_size {','.join([str(x) for x in local_size])} once bug is fixed
+    return ASTRunner(function_name, prg, global_size[::-1] if len(global_size) else [1], local_size[::-1] if len(local_size) else [1])
+
+class RawWebGPUBuffer(RawBufferCopyIn):
+  def __init__(self, size:int, dtype:DType):
+    assert dtype not in [dtypes.int8,dtypes.uint8,dtypes.int64,dtypes.uint64,dtypes.float64], f"dtype {dtype} not supported on WEBGPU"
```
Comment (jumbojets) [inline tinygrad/runtime/ops_webgpu.py:127]: could you use arbitrary float https://www.w3.org/TR/WGSL/#example-bbb80169 for float64?

**Comment by jumbojets** [comment]
Could float64 and float4 could both supported via [arbitrary float](https://www.w3.org/TR/WGSL/#example-bbb80169) and [`vec4f`](https://www.w3.org/TR/WGSL/#vec4-builtin) respectively?

**Comment by jumbojets** [comment]
Could float64 and float4 could both supported via [arbitrary float](https://www.w3.org/TR/WGSL/#example-bbb80169) and [`vec4f`](https://www.w3.org/TR/WGSL/#vec4-builtin) respectively?

**Comment by geohot** [comment]
Woah didn't realize this was ready. Let's see

**Comment by geohot** [comment]
Is there a reason this can't use C style codegen and needs a new one?

**Comment by geohot** [comment]
This is honestly an amazing PR. Raised the bounty after the fact to $500, I think this can deliver more value than I first thought.

**Comment by dc-dc-dc** [comment]
- refactored to utilize CStyleCodegen, added some extra logic and moved some stuff around
- I created a puppeteer script that loads up the page and tests the output, but looks like the CI version of chrome is not properly creating a gpu device so will just keep the compilation and leave the puppeteer script for whenever they add support for it in CI.  
CI seems like its still having issues so I posted an issue to the `wgpu-py` repo to see if they have some remedies to help resolve

**Comment by dc-dc-dc** [comment]
switched to macos-13 and disable the conv tests

**Comment by dc-dc-dc** [comment]
forcing wgpu to use metal resolved some issues in CI, just had to disable the conv2d tests but everything else is running fine

**Comment by geohot** [comment]
Yea figured it might be annoying, just the compile test is fine for now until someone makes this work.

**Comment by geohot** [comment]
Congrats! Will merge when tests pass.

Either post PayPal e-mail here or e-mail george@tinygrad.org to claim and state what split you want.
